### PR TITLE
Pro forma rate

### PR DESCRIPTION
**Pull request by MiksuJak**

### CODE REVIEW

File: contracts/truefi2/TrueFiPool2.sol [hunk 1]
```diff
@@ -662,6 +662,19 @@ contract TrueFiPool2 is ITrueFiPool2, IPauseableContract, ERC20, UpgradeableClai
...
+     * @dev Ratio of liquid assets in the pool after lending
+     * @param amount Amount of asset being lent
+     * @return Calculated ratio in basis points
+     */
+    function proFormaLiquidRatio(uint256 amount) external override view returns (uint256) {
+        uint256 _poolValue = poolValue();
+        if (_poolValue == 0) {
+            return 0;
+        }
+        return liquidValue().sub(amount).mul(BASIS_PRECISION).div(_poolValue);
```
Comment (barrutko) [inline contracts/truefi2/TrueFiPool2.sol:675]: Maybe additional brackets for clarity? Or some other sort of refactoring, like extracting `liquidValue().sub(amount)` to helper variable?
```suggestion
        return (liquidValue().sub(amount)).mul(BASIS_PRECISION).div(_poolValue);
```

File: contracts/truefi2/TrueRateAdjuster.sol [hunk 2]
```diff
@@ -6,8 +6,9 @@ import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
 import {UpgradeableClaimable} from "../common/UpgradeableClaimable.sol";
 import {ITrueFiPool2} from "./interface/ITrueFiPool2.sol";
 import {ITimeAveragedBaseRateOracle} from "./interface/ITimeAveragedBaseRateOracle.sol";
+import {ITrueRateAdjuster} from "./interface/ITrueRateAdjuster.sol";
 
-contract TrueRateAdjuster is UpgradeableClaimable {
+contract TrueRateAdjuster is UpgradeableClaimable, ITrueRateAdjuster {
```
Comment (barrutko) [inline contracts/truefi2/TrueRateAdjuster.sol:11]: This could be a separate PR 😏

**Review by barrutko — CHANGES_REQUESTED** [review]
In `LoanFactory2` we need complete rate. So not only `utilizationAdjustmentRate`, but some sort of `proFormaRate` which will consist of `baseRate + riskPremium + proFormaUtilizationAdjustmentRate + creditScoreAdjustment`

File: contracts/truefi2/TrueRateAdjuster.sol [hunk 3]
```diff
@@ -74,10 +74,22 @@ contract TrueRateAdjuster is ITrueRateAdjuster, UpgradeableClaimable {
...
+    ) external override view returns (uint256) {
+        return combinedRate(proFormaPoolBasicRate(pool, amount), creditScoreAdjustmentRate(score));
+    }
+
     function poolBasicRate(ITrueFiPool2 pool) public override view returns (uint256) {
         return min(riskPremium.add(securedRate(pool)).add(utilizationAdjustmentRate(pool)), MAX_RATE_CAP);
     }
 
+    function proFormaPoolBasicRate(ITrueFiPool2 pool, uint256 amount) public view returns (uint256) {
+        return min(riskPremium.add(securedRate(pool)).add(proFormaUtilizationAdjustmentRate(pool, amount)), MAX_RATE_CAP);
```
Comment (barrutko) [inline contracts/truefi2/TrueRateAdjuster.sol:90]: This method does exactly the same thing, the one above does. Maybe we can extract some helper method for that?

File: contracts/truefi2/TrueFiPool2.sol [hunk 4]
```diff
@@ -662,6 +662,19 @@ contract TrueFiPool2 is ITrueFiPool2, IPauseableContract, ERC20, UpgradeableClai
         return liquidValue().mul(BASIS_PRECISION).div(_poolValue);
     }
 
+    /**
+     * @dev Ratio of liquid assets in the pool after lending
+     * @param amount Amount of asset being lent
+     * @return Calculated ratio in basis points
+     */
+    function proFormaLiquidRatio(uint256 amount) external override view returns (uint256) {
```
Comment (yuchenlintt) [inline contracts/truefi2/TrueFiPool2.sol:670]: I'll add tasks to delete the original `liquidRatio()` and `utilization()` views to avoid cluttering TFP2 with unused functions. (No action needed in this PR.)

Comment (barrutko) [reply contracts/truefi2/TrueFiPool2.sol:670]: We can always use this method if needed and just pass 0 as the amount if needed. That's a good idea.

**Review by barrutko — APPROVED** [review]

**Review by vanruch — APPROVED** [review]

**Review by yuchenlintt — APPROVED** [review]
